client/asset/dcr/dcr.go

```diff
@@ -5762,93 +5832,404 @@ func (dcr *ExchangeWallet) txDB() *btc.BadgerTxDB {
+			}
+		}
+		return
+	}
+	if v := txPaysToScriptHash(msgTx); v > 0 {
+		return &asset.WalletTransaction{
+			Type:   asset.SwapOrSend,
+			ID:     tx.TxID,
+			Amount: v,
+			Fees:   fee,
+		}, nil
+	}
+
+	// Helper function will help us identify inputs that spend P2SH contracts.
+	containsContractAtPushIndex := func(msgTx *wire.MsgTx, idx int, isContract func(contract []byte) bool) bool {
+	txinloop:
+		for _, txIn := range msgTx.TxIn {
+			// not segwit
+			const scriptVer = 0
+			tokenizer := txscript.MakeScriptTokenizer(scriptVer, txIn.SignatureScript)
+			for i := 0; i <= idx; i++ { // contract is 5th item item in redemption and 4th in refund
+				if !tokenizer.Next() {
+					continue txinloop
+				}
+			}
+			if isContract(tokenizer.Data()) {
+				return true
+			}
+		}
+		return false
+	}
+
+	// Swap redemptions and refunds
+	contractIsSwap := func(contract []byte) bool {
+		_, _, _, _, err := dexdcr.ExtractSwapDetails(contract, dcr.chainParams)
+		return err == nil
+	}
+	redeemsSwap := func(msgTx *wire.MsgTx) bool {
+		return containsContractAtPushIndex(msgTx, 4, contractIsSwap)
+	}
+	if redeemsSwap(msgTx) {
+		return &asset.WalletTransaction{
+			Type:   asset.Redeem,
+			ID:     tx.TxID,
+			Amount: totalOut + fee,
+			Fees:   fee,
+		}, nil
+	}
+	refundsSwap := func(msgTx *wire.MsgTx) bool {
+		return containsContractAtPushIndex(msgTx, 3, contractIsSwap)
+	}
+	if refundsSwap(msgTx) {
+		return &asset.WalletTransaction{
+			Type:   asset.Refund,
+			ID:     tx.TxID,
+			Amount: totalOut + fee,
+			Fees:   fee,
+		}, nil
+	}
+
+	// Bond refunds
+	redeemsBond := func(msgTx *wire.MsgTx) (bool, *asset.BondTxInfo) {
+		var bondInfo *asset.BondTxInfo
+		isBond := func(contract []byte) bool {
+			const scriptVer = 0
+			lockTime, pkHash, err := dexdcr.ExtractBondDetailsV0(scriptVer, contract)
 			if err != nil {
-				dcr.log.Errorf("Error setting last query to %d: %v", tip, err)
+				return false
+			}
+			bondInfo = &asset.BondTxInfo{
+				AccountID: []byte{}, // Could look for the bond tx to get this, I guess.
+				LockTime:  uint64(lockTime),
+				BondID:    pkHash[:],
 			}
+			return true
 		}
+		return containsContractAtPushIndex(msgTx, 2, isBond), bondInfo
+	}
+	if isBondRedemption, bondInfo := redeemsBond(msgTx); isBondRedemption {
+		return &asset.WalletTransaction{
+			Type:     asset.RedeemBond,
+			ID:       tx.TxID,
+			Amount:   totalOut,
+			Fees:     fee,
+			BondInfo: bondInfo,
+		}, nil
+	}
+
+	const scriptVersion = 0
 
-		addToHistoryIfNew := func(tx *walletjson.ListTransactionsResult, txType asset.TransactionType) {
-			txHash, err := chainhash.NewHashFromStr(tx.TxID)
+	allOutputsPayUs := func(msgTx *wire.MsgTx) bool {
+		for _, txOut := range msgTx.TxOut {
+			_, addrs := stdscript.ExtractAddrs(scriptVersion, txOut.PkScript, dcr.chainParams)
 			if err != nil {
-				dcr.log.Errorf("Error decoding txid %s: %v", tx.TxID, err)
-				return
+				dcr.log.Errorf("ExtractAddrs error: %w", err)
+				return false
 			}
```

Comment:
Theres no error to check here.

Reply:
nit: `ExtractAddrs error: %w` -> `ExtractAddrs error: %v`